driver-core/src/main/com/mongodb/internal/connection/OperationContext.java

```diff
@@ -23,12 +34,87 @@
     private final long id;
+    private final ServerDeprioritization serverDeprioritization;
 
     public OperationContext() {
         id = NEXT_ID.incrementAndGet();
+        serverDeprioritization = new ServerDeprioritization();
     }
 
     public long getId() {
         return id;
     }
+
+    /**
+     * @return The same {@link ServerDeprioritization} if called on the same {@link OperationContext}.
+     */
+    public ServerDeprioritization getServerDeprioritization() {
+        return serverDeprioritization;
+    }
+
+    public static final class ServerDeprioritization {
+        @Nullable
+        private ServerAddress candidate;
+        private final Set<ServerAddress> deprioritized;
+
+        private ServerDeprioritization() {
+            candidate = null;
+            deprioritized = new HashSet<>();
+        }
+
+        ServerSelector apply(final ServerSelector selector) {
+            return new DeprioritizingSelector(selector);
+        }
+
+        void updateCandidate(final ServerAddress serverAddress, final ClusterType clusterType) {
+            candidate = isEnabled(clusterType) ? serverAddress : null;
+        }
+
+        public void onAttemptFailure(final Throwable failure) {
+            if (candidate != null && !(failure instanceof MongoConnectionPoolClearedException)) {
+                deprioritized.add(candidate);
+            }
+            candidate = null;
+        }
+
+        private static boolean isEnabled(final ClusterType clusterType) {
+            return clusterType == ClusterType.SHARDED;
+        }
+
+        /**
+         * {@link ServerSelector} requires thread safety, but that is only because a user may specify
+         * {@link com.mongodb.connection.ClusterSettings.Builder#serverSelector(ServerSelector)},
+         * which indeed may be used concurrently. {@link DeprioritizingSelector} does not need to be thread-safe.
+         */
+        private final class DeprioritizingSelector implements ServerSelector {
+            private final ServerSelector wrapped;
+
+            private DeprioritizingSelector(final ServerSelector wrapped) {
+                this.wrapped = wrapped;
+            }
+
+            @Override
+            public List<ServerDescription> select(final ClusterDescription clusterDescription) {
+                if (isEnabled(clusterDescription.getType())) {
+                    List<ServerDescription> filteredServerDescriptions = ClusterDescriptionHelper.getServersByPredicate(
+                            clusterDescription, serverDescription -> !deprioritized.contains(serverDescription.getAddress()));
+                    ClusterDescription filteredClusterDescription = new ClusterDescription(
+                            clusterDescription.getConnectionMode(),
+                            clusterDescription.getType(),
+                            clusterDescription.getSrvResolutionException(),
+                            filteredServerDescriptions,
```

Comment:
Let's add a `with...` method.

Reply:
`ClusterDescription` is part of the API, and so far each time we needed more fields in such a class, we have been adding a new constructor, sometimes deprecating the previous constructors. One can certainly propose us to change that API approach not for just this class, but for all similar situations, but that warrants a team discussion. Since this PR does not add a new constructor to `ClusterDescription`, and only uses an existing constructor, it is not blocked by the decision on such a change, were it to be proposed.

Reply:
This can be a private static method.

Reply:
[Done](https://github.com/mongodb/mongo-java-driver/pull/1367/commits/6574e241434f8c8946984ade559695e2aa9f29b3), only in Java SE 8 we are not allowed to have `static` methods in inner classes. This is an artificial limitation (I guess, we are not the only ones who don't like to complicate things), which was removed in Java SE 16 (see https://openjdk.org/jeps/395#Static-members-of-inner-classes, https://bugs.openjdk.org/browse/JDK-8254321).